### PR TITLE
Update _dedup_save_plans.py

### LINKED ISSUE #125740: [Distributed Checkpoint] When loading FSDP sharded checkpointing  each rank needs all the checkpointing files
### 🐛 Describe the bug

Hi team, when loading FSDP sharded checkpoint without changing the world size (no resharding), each rank needs to get checkpoint file saved by other ranks.  This happens when there are scalar tensors in the optimizer. 

Below is a simple two-gpu reproduce:

```
import os
import torch
import torch.distributed as dist
import torch.multiprocessing as mp
import torch.nn as nn
from typing import Callable
from torch.distributed import checkpoint as dist_cp
from torch.distributed.checkpoint.planner import LoadItemType
from torch.distributed.checkpoint.state_dict import get_state_dict, StateDictOptions
from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
from torch.distributed.fsdp.api import ShardingStrategy
from torch.optim import SGD

class MyModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.linear_0 = nn.Linear(3, 3)
        self.linear_1 = nn.Linear(3, 3)
    def forward(self, x):
        return torch.sum(self.linear_1(self.linear_0(x)))

class MyReader(dist_cp.FileSystemReader):
    def read_data(self, plan, planner):
        need_other_rank_file = False
        for read_item in plan.items:
            relative_file_path = self.storage_data[read_item.storage_index].relative_path
            rank = torch.distributed.get_rank()
            if read_item.type == LoadItemType.TENSOR:
                if (rank == 0 and "__1_0" in relative_file_path) or (rank == 1 and "__0_0" in relative_file_path):
                    print(f"bigning debug rank: {torch.distributed.get_rank()}, path: {relative_file_path}, {read_item}")
                    need_other_rank_file = True
        if need_other_rank_file:
            pass
            #raise RuntimeError("Why rank 0 needs '__1_0.distcp' ?")
        return super().read_data(plan, planner)

class MySGD(SGD):
    def __init__(
        self,
        params,
        lr,
        momentum: float = 0,
        dampening: float = 0,
        weight_decay: float = 0,
        nesterov: bool = False,
    ):
        super().__init__(
            params=params,
            lr=lr,
            momentum=momentum,
            dampening=dampening,
            weight_decay=weight_decay,
            nesterov=nesterov,
        )

    @torch.no_grad()
    def step(self, closure=None):
        for group in self.param_groups:
            for p in group['params']:
                state = self.state[p]
                if 'step' not in state:
                    state['step'] = torch.zeros((), dtype=torch.float, device=p.device)
                state['step'] += 1
        super().step(closure)
    
def main(rank, world_size):
    ## INITIALIZE DIST
    # Running on one node so master_addr is just local host
    os.environ["MASTER_ADDR"] = "localhost"
    os.environ["MASTER_PORT"] = "28000"
    # All ranks simulataneously init the process group together.
    dist.init_process_group("nccl", rank=rank, world_size=world_size)
    torch.cuda.set_device(rank)

    save_path = "./checkpoint"

    model = MyModel().to(f"cuda:{rank}")
    optimizer = MySGD(model.parameters(), lr=0.01)
    fsdp_model = FSDP(
        model,
        sharding_strategy=ShardingStrategy.FULL_SHARD,
        use_orig_params=True,
    )
    optimizer.zero_grad()

    x = torch.rand(([2, 3]), device=f"cuda:{rank}")
    y = fsdp_model(x)
    y.backward()
    optimizer.step()

    # save model
    model_state_dict, opt_state_dict = get_state_dict(
        fsdp_model, 
        optimizer, 
        options=StateDictOptions(full_state_dict=False),
    )
    state_dict = {
        "model": model_state_dict,
        "optimizer": opt_state_dict,
    }
    dist_cp.save(
        state_dict=state_dict,
        storage_writer=dist_cp.FileSystemWriter(save_path)
    )
    print(f"bigning debug saving done")

    # load
    dist_cp.load(
        state_dict=state_dict,
        storage_reader=MyReader(save_path),
    )

    dist.destroy_process_group()

if __name__ == "__main__":
    world_size = 2 
    mp.spawn(
        main,
        args = (world_size, ),
        nprocs=world_size,
        join=True
    )
```
It printed that rank 0 read_items are in `__1_0.distcp`.  output screenshot: 
image

so some of the scalar tensors needed by rank0 are only in `__1_0.distcp` file, which is saved by rank 1.  Imagine for a multiple node training, when loading checkpoint, node-0 rank needs files saved by other nodes which are **not** available on node-0. Or with remote checkpointing saved in cloud, **each node needs to download all the sharded checkpoint files**. 

I think it's because here when deduping save plans , [it tries to balance the storage needed for each rank, so those replicated scalar tensors are saved in to different files](https://github.com/pytorch/pytorch/blob/main/torch/distributed/checkpoint/_dedup_save_plans.py#L28-L32). So when loading the checkpoint, even without resharding, each rank still needs almost all files to load the tensor. Can we save those replicate tensors to rank-0 only ?

### Versions

PyTorch version: 2.3.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: Could not collect
CMake version: version 3.26.3
Libc version: glibc-2.31

Python version: 3.11.9 (main, Apr  6 2024, 17:59:24) [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-91-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A100-SXM4-40GB
GPU 1: NVIDIA A100-SXM4-40GB

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### PR DESCRIPTION
**Pull request by bigning**
To resolve https://github.com/pytorch/pytorch/issues/125740, save each tensor on the lowest rank.

Fixes #125740

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @LucasLLC

### CODE REVIEW

**Review by LucasLLC — CHANGES_REQUESTED** [review]
At minimum, we would need this to be optional and it should not replace the current deduplication logic. The current logic is a storage optimization, and this would cause a regression in terms of calling dcp.save on models which are replica heavy.

I think in the originally linked issue, we considered not applying this logic to scalars as a fix for minimizing the number of files which need to be loaded

**Comment by bigning** [comment]
>  The current logic is a storage optimization, and this would cause a regression in terms of calling dcp.save on models which are replica heavy.

it's "optimization" only in terms of saving balance. But it hurts the loading performance in multi-node case. 

> we considered not applying this logic to scalars as a fix

i don't think that works. The root cause is the duplicated tensors are saved in different files, no matter if it's scalar tensor or not.

**Comment by bigning** [comment]
@LucasLLC , i replied to your comment. Could you please take a look ?

**Comment by LucasLLC** [comment]
@bigning , I believe this generally this isn't seen as a large issue during saving since files are all expected to live in the same NFS directory. Could we change this PR to make the de-duplication optional (and true by default)?

File: torch/distributed/checkpoint/_dedup_save_plans.py [hunk 1]
```diff
@@ -14,27 +14,20 @@
...
     to_remove: List[Set] = [set() for _ in range(len(all_plans))]
-    plan_to_size = [0] * len(all_plans)
     for write_item_idx, plan_indices in write_item_to_plan_indices.items():
-        select_plan_idx = min(plan_indices, key=lambda plan_idx: plan_to_size[plan_idx])
-
-        write_item = write_item_idx_to_write_item[write_item_idx]
-        # essentially ignores the storage size of anything that is not a tensor, since
-        # we don't know how much storage they represent
-        plan_to_size[select_plan_idx] += write_item.tensor_storage_size() or 1
+        select_plan_idx = min(plan_indices, key=lambda plan_idx: plan_idx)
```
Comment (Skylion007) [inline torch/distributed/checkpoint/_dedup_save_plans.py:30]:
```suggestion
        select_plan_idx = min(plan_indices)
```

**Comment by bigning** [comment]
> make the de-duplication optional 

If we skip the dedup, it fails here https://github.com/pytorch/pytorch/blob/main/torch/distributed/checkpoint/default_planner.py#L386-L387. 

> files are all expected to live in the same NFS directory.

For NFS, the issue is not about if the file exists in the NFS. It's that each node needs to download the Nx times more files. If it's using cloud, this introduces:
1. Nx cloud storage money cost for downloading more data
2. Nx more time blocking the training. For saving, it can be async, but for loading, usually loading checkpointing blocks the training. 
3. Downloading large files from cloud is not stable, this is really a pain point for using cloud storage. The more files you download, the more chance the checkpointing downloading fails. 

Can I just add a `save_to_same_rank`  param to `dedup_save_plans` and `DefaultSavePlanner` ?

**Comment by LucasLLC** [comment]
> Can I just add a `save_to_same_rank` param to `dedup_save_plans` and `DefaultSavePlanner`

@bigning I understand the pain points in having to download multiple files. I think we would accept a PR which makes this behavior configurable.

**Comment by bigning** [comment]
> makes this behavior configurable.

@LucasLLC , this makes sense. I added a param to `dedup_save_plans` and `DefaultSavePlanner`, could you take a look? Thanks!

**Comment by LucasLLC** [comment]
Thanks @bigning ! This looks good to me. Will merge if tests pass

**Review by LucasLLC — APPROVED** [review]

**Comment by bigning** [comment]
thanks @LucasLLC , it seems two lints failed, I can't find useful error message, do you know how to fix or how to re-run?

**Comment by bigning** [comment]
> thanks @LucasLLC , it seems two lints failed, I can't find useful error message, do you know how to fix or how to re-run?

NVM, i just submitted another commit. Looks all tests are green now.

**Comment by johnmullan** [comment]
@bigning would this PR allow me to pull down specific `__{rank}_0.distcp` files from cloud storage onto a local NVME (provided world size stays same) and load from there? For more context I'm trying to prevent loading from an NFS to improve start up times.

**Comment by bigning** [comment]
yeah, this PR allows you to download as few files as possible for each rank. Before this PR, if you have non-sharded tensors in the state dict that are needed by each rank, those tensors would be saved to different files. So each rank needs to download many files. After the PR, those non-sharded tensors are only saved to one file.

**Pull request closed without merging**